containerd-shim-v2/service.go

```diff
@@ -550,7 +550,35 @@ func (s *service) Resume(ctx context.Context, r *taskAPI.ResumeRequest) (*ptypes
 
 // Kill a process with the provided signal
 func (s *service) Kill(ctx context.Context, r *taskAPI.KillRequest) (*ptypes.Empty, error) {
-	return nil, errdefs.ErrNotImplemented
+	s.Lock()
+	defer s.Unlock()
+
+	signum := syscall.Signal(r.Signal)
+
+	c, err := s.getContainer(r.ID)
+	if err != nil {
+		return nil, err
+	}
+
+	processID := c.id
+	if r.ExecID != "" {
+		execs, err := c.getExec(r.ExecID)
+		if err != nil {
+			return nil, err
+		}
+		processID = execs.id
+	}
+
+	err = s.sandbox.SignalProcess(c.id, processID, signum, r.All)
+	if err != nil {
+		return nil, err
+	}
+
+	if signum == syscall.SIGTERM {
+		err = s.sandbox.SignalProcess(c.id, processID, syscall.SIGKILL, r.All)
```

Comment:
Why are we sending a `SIGKILL` here? The signal is a `SIGTERM` and has already been sent.
If there's a good explanation for this, please add a comment in the code explaining this.

Reply:
Since kubernetes will use SIGTERM signal to stop a container by default, but some container process such as "shell" will not response to this signal, thus container stop will always get timeout response, in order to terminate the container successfully, here send another SIGKILL to make sure the process will be terminated successfully.  I'll add it to comments.

Reply:
Well I understand but I would expect the CRI layer (CRI-containerd or CRI-O) to actually send a SIGKILL themselves if they realize that SIGTERM didn't work. Is it something that we want to handle at this level?
/cc @bergwolf ^

Reply:
I think we should follow runc's practice and only send `SIGTERM`. If the container process chooses to ignore `SIGTERM`, the upper layer should send other signals after timeout. @lifupan do you see any test cases that runc passes but shimv2 fails due to `SIGTERM` timeout?

Reply:
@bergwolf exactly, the upper layer knows about the workload and should be the one that knows if SIGTERM is appropriate or not.

Reply:
@bergwolf @sboeuf  The reason I added this KILL signal because for kata cli it also will send another KILL signal to make sure stop the container once it received SIGTERM and SIGKILL signal.

Reply:
@lifupan let's not duplicate some fix that we had to use with kata CLI since they might not be needed in this case. Do you have a concrete test case that will actually fail if you don't send a SIGKILL after a SIGTERM ?

Reply:
Yes, IIRC the interactive shell ignores `SIGTERM`, I think it would respond to the `SIGTERM` if you run without `-it`.

Reply:
Fair enough, it's not ideal, but we'll fix that later once we understand better the issue. I don't want to delay this PR because of this detail.

Reply:
Thanks @sboeuf @jodh-intel @WeiZhang555 @bergwolf  @egernst. Thanks for your patient and careful help and review.